cvmfs/catalog_mgr.h

```diff
@@ -53,7 +53,113 @@ enum LoadError {
   kLoadNumEntries
 };
 
-inline const char *Code2Ascii(const LoadError error) {
+/**
+ * Location of the most recent root catalog.
+ * Used as part of the process of loading a catalog.
+ * - GetNewRootCatalogContext() sets the location within the CatalogContext obj
+ * - LoadCatalogByHash(): when loading a root catalog it uses the location
+ *                        stored within the CatalogContext object to retrieve
+ *                        the root catalog from the right location
+ */
+enum RootCatalogLocation {
+  kCtlogLocationUnknown = 0,
+  kCtlgLocationMounted,      // already loaded in mounted_catalogs_
+  kCtlgLocationServer,
+  kCtlgLocationBreadcrumb
+};
+
+/**
+ * CatalogContext class contains all necessary information to load a catalog and
+ * also keeps track of the resulting output.
+ * It works the following:
+ * 1) Load a new root catalog:
+ *  - Use empty constructor CatalogContext()
+ *  - Let the CatalogContext object be populated by GetNewRootCatalogContext()
+ *    - This will set: hash, mountpoint, root_ctlg_revision, root_ctlg_location
+ *  - Call LoadCatalogByHash()
+ *    - This will set: sqlite_path
+ * 2) Load a catalog based on a given hash
+ *  - Populate CatalogContext object; used constructor depends on catalog type
+ *    - Root catalog: CatalogContext(shash::Any hash, PathString mountpoint,
+              RootCatalogLocation location)
+      - Nested catalog: CatalogContext(shash::Any hash, PathString mountpoint)
+      - Note: in this case root_ctlg_revision is not used
+ *  - Call LoadCatalogByHash()
+      - This will set: sqlite_path
+ */
+struct CatalogContext {
+ public:
+  CatalogContext() :
+              hash_(shash::Any()),
+              mountpoint_(PathString("invalid", 7)),
+              sqlite_path_(""),
+              root_ctlg_revision_(-1ul),
+              root_ctlg_location_(kCtlogLocationUnknown),
+              manifest_ensemble_(NULL) { }
+  CatalogContext(shash::Any hash, const PathString &mountpoint) :
+              hash_(hash),
+              mountpoint_(mountpoint),
+              sqlite_path_(""),
+              root_ctlg_revision_(-1ul),
+              root_ctlg_location_(kCtlogLocationUnknown),
+              manifest_ensemble_(NULL) { }
+
+  CatalogContext(shash::Any hash, const PathString &mountpoint,
+              RootCatalogLocation location) :
+              hash_(hash),
+              mountpoint_(mountpoint),
```

Comment:
Wouldn't the mountpoint always be empty?
```suggestion
  CatalogContext(const shash::Any &hash, RootCatalogLocation location) :
              hash_(hash),
              mountpoint_(),
```

Reply:
no there are cases where we prefill the mountpoint that does not have to be the root catalog. e.g. in `MountCatalog` or in `LoadFreeCatalog`

Reply:
But if it is not a root catalog we are loading, why would we use the constructor that takes a RootCatalogLocation?